### PR TITLE
Adding djinn and tonic files

### PR DESCRIPTION
**Pull request by Lilisys**
Restructured the directories to include
- gen-cpp - the files generated by thrift that are common to every IPAService
- djinn - contains a cc and non-cc version. The non-cc version can be used with tonic-img
- tonic-common - files needed by tonic-img and tonic-nlp (like weights and configs)

### CODE REVIEW

**Comment by Lilisys** [comment]
The existing files can run with or without the command center. Here's what still needs to be done:
- [ ] Write a readme
- [ ] Incorporate Tim's structural changes for standalone version
- [ ] Incorporate Tim's structural changes for cc version
- [ ] rebase this

File: gen-cpp/commons.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
+        SchedulerServiceClient *client = new SchedulerServiceClient(protocol);
+        transport->open();
+        return client;
+}
+	
+	TServers::TServers() {
+	}
+void serverThread(int port, boost::shared_ptr<TProcessor> &processor) {
+	// boost::shared_ptr<TProtocolFactory> protocolFactory(new TBinaryProtocolFactory());
```
Comment (jhauswald) [inline gen-cpp/commons.cpp]: why is there commented out code?

File: djinn/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+###Running Djinn###
+In a separate terminal, run an instance of the command center: _This section is still under construction_
+
+In this terminal, run these commands from the `djinn/` directory:
+```
+$ make
+$ ./DjinnService
+```
+
+Now you can run services such as [Tonic Image Processing](https://github.com/Lilisys/clarityeco/tree/imgMerge/tonic-img) or [Tonic Natural Language Processing](https://github.com/Lilisys/clarityeco/tree/nlpMerge).
```
Comment (jhauswald) [inline djinn/README.md]: Your Readme's are pointing to your repo, we'll have to make sure they're correctly pointing when this gets merge.

File: djinn/DjinnService.cpp [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+class DjinnServiceHandler : virtual public IPAServiceIf {
+ public:
+  DjinnServiceHandler() {
+    // Your initialization goes here
+    this->SERVICE_NAME = SERVICE_TYPE;
+    this->SCHEDULER_IP = "141.212.106.68";
+    this->SCHEDULER_PORT = 8888;
+    this->SERVICE_IP = "clarity07.eecs.umich.edu";
+    this->SERVICE_PORT = 7071;
```
Comment (jhauswald) [inline djinn/DjinnService.cpp]: I know these are all just defaults that eventually get set, but probably not a good idea to have any real IPs, let use `localhost:8000` and `localhost:8080` as defaults for scheduler and service for example. Same thing in the parse_opts function.

**Comment by jhauswald** [comment]
**Observations:**
- There is a mix of `cout`, `LOG(INFO)` and even a `printf`, choose one (preferably glog since it gives a timestamp and location of where the message was printed from.
- How expensive is the lexical cast?
- Let's get rid of the warmup pass on the GPU, this was there for experiments but in this case djiNN is running on loop. What happens here is the `.caffemodel` gets moved to the GPU during the first query and then stays there until djinn closes this is why the first query is going to be substantially slower. The scheduler should have logic to throw this run out if its making decisions based on this.
- So does djinn send the command-center the latency of a forward pass every time? Can you elaborate on this?
- This isn't WIP anymore, you've made the changes to be sending the real image over right?

**Enhancement:**
Can you test djinn with the latest version of Caffe (clone and build it from the original repo)? We'd like to have djinn be compatible across the versions. Some networks may not work (FACE and NLP ones) because they don't have the layers, focus on just IMC for example. [This](https://github.com/Lilisys/clarityeco/blob/mergeTonic/djinn/DjinnService.cpp#L116) line will most definitely break, [this](https://github.com/ypkang/djinn/blob/mobile/dnn/src/main.cpp#L114) is how its set now (see the slight difference?).
if you can figure out a good way to check the versions then use we can have support for the old caffe and the newest version(s).

**Comment by Lilisys** [comment]
Observations:
1. I'll change them all to glog
2. [this](http://www.boost.org/doc/libs/1_49_0/doc/html/boost_lexical_cast/performance.html) page compares the performance of lexical cast to stringstreams and printf/scanf. gcc 4.8 isn't on there, but the general trend is that string->float is fastest with lexical_cast while float->string is fastest with scanf/printf. The reason lexical cast is being used is because previously I had tried memcopy or memove, both of which kept segfaulting. 
3. sounds good
4. Are you looking at the `PRINT_STAT_DOUBLE("inference latency", toc());` line? Or the section in `submitQuery`? The former comes from the old code, while the latter comes from @Thomas-Yang's command center structure. You'll have to ask him for details on this. 
5. Yep, I'll change the name

Enhancement:
I'll add this to my todo list for the week. I'll document issues I run into as well as the general process in the wiki.

File: djinn/DjinnService.cpp (hunk 3 above)
Comment (Lilisys) [inline djinn/DjinnService.cpp]: So would it look like this?
```
 DjinnServiceHandler() {
    this->SERVICE_NAME = SERVICE_TYPE;
    this->SCHEDULER_IP = "localhost";
    this->SCHEDULER_PORT = localhost:8000;
    this->SERVICE_IP = "localhost";
    this->SERVICE_PORT = localhost:8080;
}
```
And are we setting every service default to 8080? Or just keeping the various existing defaults?

**Comment by jhauswald** [comment]
Here's what I tried:
- Running Thomas-yang/service_project@51dff91cf7ed7fff87cd575b05dfea30e89429d4 at `localhost:8000`
- check'ed out this branch everything compiled fine
- when I try to run djinn with `./DjinnService --ccip localhost --ccport 8080 -p 8080` I get:
```
I0803 22:58:25.612004 24275 DjinnService.cpp:140] registering to command center running at localhost:8080
terminate called after throwing an instance of 'apache::thrift::TApplicationException'
  what():  Invalid method name: 'registerBackend'
Aborted (core dumped)
```

Seems there's a version issue with the command center. If I could get a commit hash, I can build it but currently a little stuck. We can figure this out tomorrow and you can walk me through how you're running it. I'd like to have the command center running, checkout this branch, have djinn running, checkout #40 and execute end-to-end. Once all that is merge, should be straightforward for me to test #41 .

**Comment by jhauswald** [comment]
Scratch that, I'm an idiot, let me keep going...

**Comment by jhauswald** [comment]
Ok I tested this with #41. It works and we've been through the style so I just tested correctness so merging.

**Comment by jhauswald** [comment]
This is requesting to merge to `master`, the PR needs to go to `dev`. I reverted the PR for now, @Lilisys can you make it to dev and i'll merge it?

**Comment by Lilisys** [comment]
@jhauswald new one is #50

**Pull request merged**